mule-extensions-api/src/main/java/org/mule/runtime/extension/api/OnTerminateResult.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.extension.api;
+
+import org.mule.runtime.api.message.Error;
+import org.mule.runtime.extension.api.annotation.execution.OnTerminate;
+
+import java.util.function.Consumer;
+
+/**
+ * Result for the {@link OnTerminate} callbacks.
+ * This result provides an easy way and common way to discern the different kinds of states that a
+ *
+ * @since 1.0
+ */
+@FunctionalInterface
+public interface OnTerminateResult {
+
+  void execute(Runnable onSuccess, Consumer<Error> onParameterResolutionError, Consumer<Error> onResponseError);
```

Comment:
This API is:

* Not clear
* Inefficient to use, since it always forces at least one empty lambda
* Doesn't support all possible applications of the onTerminate
* Leads to weird onTerminate() signatures, why should the user receive both a SourceCallbackContext and one of these?

This should be reworked as one single object which contains:

Optional<Message> resultMessage
Optional<Error> onParameterResolutionError
Optional<Error> onResponseError